### PR TITLE
[FG:InPlacePodVerticalScaling] Rework handling of allocated resources

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Improve the consistency of handling allocated vs. requested resources.

The previous implementation stored allocated resources in the pod status, but downstream consumers were inconsistent about when the spec resources vs allocated resources were used. Furthermore, it is often unclear from Kubelet code what version the `pod.Status` represents, and how up-to-date it is.

This change handles the resize logic at the beginning of the pod sync loop, and overwrites the pod spec with the allocated resources as necessary. As a result, downstream consumers do not need to worry about which resources to pull from (spec vs allocated) for the most part.

#### Which issue(s) this PR fixes:
For https://github.com/kubernetes/kubernetes/issues/128065
Related to https://github.com/kubernetes/kubernetes/issues/116971

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig node
/priority important-soon
/milestone v1.32

### CODE REVIEW

**Comment by knabben** [comment]
Hello @tallclair 
The code freeze is starting 02:00 UTC Friday November 8th 2024 (about 2 weeks from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time. Just a gently reminder of the date since PR is progressing.
Thank you!

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2296,9 +2296,7 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 		// Use allocated resources values from checkpoint store (source of truth) to determine fit
 		otherPods := make([]*v1.Pod, 0, len(pods))
 		for _, p := range pods {
-			op := p.DeepCopy()
-			kl.updateContainerResourceAllocation(op)
-
+			op, _ := kl.statusManager.UpdatePodFromAllocation(p)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:2299]: hmmm....could we add a comment in `canAdmitPod` that it assumes the pod spec contains the allocated resources? Right now the `pods` argument takes the vanilla pod specs, while `pod` takes the spec with allocated resources. 

Another option is to change the arg name to reflect this:
```
canAdmitPod(pods []*v1.Pod, allocatedPod *v1.Pod)
```

And I know this is not the ideal state that we'd like, but making it more readable would help for now.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2749,36 +2735,47 @@ func (kl *Kubelet) HandlePodSyncs(pods []*v1.Pod) {
...
 			}
 		}
 	}
 	return false
 }
 
-func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, *v1.Pod, v1.PodResizeStatus) {
+// canResizePod determines if the requested resize is currently feasible.
+// Returns true if the resize can proceed.
+func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, v1.PodResizeStatus) {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:2764]: A similar comment here. Could we make it clear that "allocatedPod" is being used here?

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -2790,76 +2787,42 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, *v1.Pod, v1.PodResizeStatus)
...
 		// Update pod resource allocation checkpoint
-		if err := kl.statusManager.SetPodAllocation(updatedPod); err != nil {
-			//TODO(vinaykul,InPlacePodVerticalScaling): Can we recover from this in some way? Investigate
-			klog.ErrorS(err, "SetPodAllocation failed", "pod", klog.KObj(updatedPod))
-			return pod
+		if err := kl.statusManager.SetPodAllocation(pod); err != nil {
+			return nil, err
 		}
+	} else {
+		// If resize isn't immediately feasible, proceed with the allocated pod.
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:2815]: if resizing is feasible, the allocated pod would be the same as the pod, no? Why separate the two?

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1975,7 +1975,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 		return false, nil
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeInProgress(pod, &apiPodStatus) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeInProgress(pod, podStatus) {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:1978]: Question about the comment below: why do we need to periodically call PLEG to update the cache. I'm not familiar with the original implementation of pod resizing, so likely I missed the point here.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -1789,6 +1788,17 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 		}
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && !kubetypes.IsStaticPod(pod) {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:1791]: There is a function `isInPlacePodVerticalScalingAllowed` that basically checks the same thing. Can we just reuse it?

File: pkg/kubelet/qos/policy.go [hunk 6]
```diff
@@ -63,11 +60,6 @@ func GetContainerOOMScoreAdjust(pod *v1.Pod, container *v1.Container, memoryCapa
 	// targets for OOM kills.
 	// Note that this is a heuristic, it won't work if a container has many small processes.
 	memoryRequest := container.Resources.Requests.Memory().Value()
-	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```
Comment (yujuhong) [inline pkg/kubelet/qos/policy.go:66]: Could you add a comment to the function stating that we calculate the oom score based on the allocated resources (as opposed to the desired resources in the spec)?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:2299]: I actually went the other direction, and moved the updating of `otherPods` out of this function, since now we can just use the `getAllocatedPods()` function, and renamed the pods argument accordingly.

The `pod` argument is actually usually _not_ the allocated pod, since we want to check the desired resources against admission to see if they can be allocated. Basically, there are 3 ways this function can be called:
1. New pod scheduled to this node: no allocation yet, pod is the desired pod
2. Resized pod: pod has an allocation, but we want to see if we can admit the new desired state
3. Kubelet restart, re-admitting existing pods: pod has an allocation; even if there's a pending resize, we want to re-admit the allocated state first to make sure we don't reject a pod that's already running. This is the only case where this function is called with an allocated pod.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:2764]: This should **not** be an allocated pod.

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:2815]: I'm not sure I totally understand what you're suggesting, but I flipped this around to hopefully make it more clear. If the pod doesn't fit, we should proceed with the allocated pod. If it does fit, then we set the allocation to be the new state. Is that what you meant?

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go [hunk 7]
```diff
@@ -2188,18 +2189,19 @@ func TestComputePodActionsForPodResize(t *testing.T) {
...
 	for desc, test := range map[string]struct {
-		podResizePolicyFn       func(*v1.Pod)
-		mutatePodFn             func(*v1.Pod)
+		setupFn                 func(*v1.Pod, *kubecontainer.PodStatus)
 		getExpectedPodActionsFn func(*v1.Pod, *kubecontainer.PodStatus) *podActions
 	}{
 		"Update container CPU and memory resources": {
-			mutatePodFn: func(pod *v1.Pod) {
-				pod.Spec.Containers[1].Resources = v1.ResourceRequirements{
+			setupFn: func(pod *v1.Pod, status *kubecontainer.PodStatus) {
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:2196]: not exactly. since we're no longer fetching the actual resources from `pod.Status`, we need to configure the `kubecontainer.PodStatus` here, so I renamed the function accordingly.

Collapsing `podResizePolicyFn` into `setupFn` is just cleanup. I don't know why that had its own function. This test was a bit of a mess, TBH.

**Review by tallclair — COMMENTED** [review]
Thanks for the review @yujuhong !

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (vinaykul) [reply pkg/kubelet/kubelet.go:1982]: Thanks for adding the comment I had missed (important in hingsight): https://github.com/kubernetes/kubernetes/pull/102884#discussion_r817178710 . Eventually I hope EventedPLEG will take care of this and replace the hack.

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -1789,6 +1789,20 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 		}
 	}
 
+	// handlePodResourcesResize updates the pod to use the allocated resources. This should come
+	// before the main business logic of SyncPod, so that a consistent view of the pod is used
+	// across the sync loop.
+	if kuberuntime.IsInPlacePodVerticalScalingAllowed(pod) {
+		// Handle pod resize here instead of doing it in HandlePodUpdates because
```
Comment (vinaykul) [inline pkg/kubelet/kubelet.go:1796]: Do we lose anything by dropping the check for CouldHaveRunningContainers [below](https://github.com/kubernetes/kubernetes/pull/128269/files#diff-67dd9e8f3ee257072765326cb4f242852554a2c0753563fa51e292c0a63a7b94L1951)? (I have to take a close look at the updated handlePodResize logic tomorrow, but just confirming this is intentional)

File: pkg/kubelet/status/status_manager.go [hunk 9]
```diff
@@ -255,10 +298,7 @@ func (m *manager) SetPodAllocation(pod *v1.Pod) error {
 	m.podStatusesLock.RLock()
 	defer m.podStatusesLock.RUnlock()
 	for _, container := range pod.Spec.Containers {
-		var alloc v1.ResourceList
-		if container.Resources.Requests != nil {
-			alloc = container.Resources.Requests.DeepCopy()
-		}
+		alloc := *container.Resources.DeepCopy()
 		if err := m.state.SetContainerResourceAllocation(string(pod.UID), container.Name, alloc); err != nil {
```
Comment (vinaykul) [inline pkg/kubelet/status/status_manager.go:302]: Rename? SetContainerResourceAllocation -> SetContainerResources.

File: pkg/kubelet/status/status_manager.go [hunk 10]
```diff
@@ -236,12 +246,45 @@ func (m *manager) Start() {
 
 // GetContainerResourceAllocation returns the last checkpointed AllocatedResources values
 // If checkpoint manager has not been initialized, it returns nil, false
-func (m *manager) GetContainerResourceAllocation(podUID string, containerName string) (v1.ResourceList, bool) {
+func (m *manager) GetContainerResourceAllocation(podUID string, containerName string) (v1.ResourceRequirements, bool) {
```
Comment (vinaykul) [inline pkg/kubelet/status/status_manager.go:249]: Likewise, rename GetContainerResourceAllocation -> GetContainerResources?

File: pkg/kubelet/status/fake_status_manager.go [hunk 11]
```diff
@@ -63,31 +63,34 @@ func (m *fakeManager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
...
+	allocs := m.state.GetPodResourceAllocation()
+	return updatePodFromAllocation(pod, allocs)
 }
 
 func (m *fakeManager) SetPodAllocation(pod *v1.Pod) error {
 	klog.InfoS("SetPodAllocation()")
 	for _, container := range pod.Spec.Containers {
-		var alloc v1.ResourceList
+		var alloc v1.ResourceRequirements
 		if container.Resources.Requests != nil {
```
Comment (vinaykul) [inline pkg/kubelet/status/fake_status_manager.go:84]: Do we need this nil check anymore?

File: pkg/kubelet/kubelet.go (hunk 8 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:1796]: I think the check was needed before because `AllocatedResources` was only set in the container status for running containers. Now that we're pulling from the internal store directly, this is no longer needed. Removing this allows pods to be resized even with no containers running, which can also reduce the restarts or resize operations needed.

File: pkg/kubelet/status/status_manager.go (hunk 9 above)
Comment (tallclair) [reply pkg/kubelet/status/status_manager.go:302]: I'm planning a follow-up PR to refactor the allocation logic out of status manager to something like a new "allocation manager". Mind if I defer this to that PR?

File: pkg/kubelet/status/status_manager.go (hunk 10 above)
Comment (tallclair) [reply pkg/kubelet/status/status_manager.go:249]: Same as above.

File: pkg/kubelet/status/fake_status_manager.go (hunk 11 above)
Comment (tallclair) [reply pkg/kubelet/status/fake_status_manager.go:84]: Good catch. Fixed.

File: pkg/kubelet/kubelet.go [hunk 12]
```diff
@@ -2754,117 +2735,87 @@ func (kl *Kubelet) HandlePodSyncs(pods []*v1.Pod) {
...
 	memAvailable := node.Status.Allocatable.Memory().Value()
-	cpuRequests := resource.GetResourceRequest(podCopy, v1.ResourceCPU)
-	memRequests := resource.GetResourceRequest(podCopy, v1.ResourceMemory)
+	cpuRequests := resource.GetResourceRequest(pod, v1.ResourceCPU)
+	memRequests := resource.GetResourceRequest(pod, v1.ResourceMemory)
 	if cpuRequests > cpuAvailable || memRequests > memAvailable {
-		klog.V(3).InfoS("Resize is not feasible as request exceeds allocatable node resources", "pod", podCopy.Name)
-		return false, podCopy, v1.PodResizeStatusInfeasible
+		klog.V(3).InfoS("Resize is not feasible as request exceeds allocatable node resources", "pod", klog.KObj(pod))
+		return false, v1.PodResizeStatusInfeasible
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubelet/kubelet.go:2777]: I think you're referring to https://github.com/kubernetes/kubernetes/issues/114203? I think it's a good idea, and definitely simplifies some of the infeasible resource handling, but I'm not sure it needs to block beta.

> Is there a use-case where only node can evaluate infeasibility outcome for CPU/memory?

For now, we're going to disallow resizing guaranteed pods on nodes with a static CPU policy. I assume this will be implemented as marking the resize as infeasible. (This is a temporary change, just removing static CPU policy from alpha scope. It will have a separate feature gate.)

**Comment by tallclair** [comment]
Thanks for the review, @vinaykul !

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet.go]: Enumerating the cases where the function is called is actually very helpful. Verifying my understanding below:
 - (1) and (3). Called in `HandlePodAddition` -- we look at the pod and its allocated resources (if it exists) and the desired spec (if it's considered new).
 - (2). Called in `canResizePod` -- we look at the pod's updated desired spec

Does that sound right to you?

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet.go]: My mental model (which might be wrong):

1. If resizing is feasible, update the allocated resources of the pod
2. If resizing is not feasible, the allocated resources stays the same

Either way, we'd be looking at "(updated) allocated resources" of the pod. I think the naming of the variables here might be my source of confusion....


Another question I have is that do we always set "allocated resources" even when there's no resize request? If not, doing that might help simplify some logic in the code.

**Review by yujuhong — COMMENTED** [review]
Looks good overall. Have some minor questions

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go]: Yes, correct.

I'd add this as a comment, but I'm not sure where. I don't really like add comments about callers to functions, because in my experience they tend to be hard to keep up-to-date.

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go]: Yeah. I updated more references to be to the `allocatedPod`, hopefully it's more clear now.

> Another question I have is that do we always set "allocated resources" even when there's no resize request? If not, doing that might help simplify some logic in the code.

The first part of the function gets the allocated pod, but if there's no resize (no update), it just returns early with that. Is that what you mean?

Would it help to rename this function? Maybe `handlePodAllocation` instead?

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:1982]: https://github.com/kubernetes/kubernetes/pull/123941 is attempting to fix this, but I'm skeptical that that approach is feasible. I think we have 2 options:
1. Augment the CRI API, either to include resources in ListContainers, or add something like ListContainerStatuses
2. Add some way to indicate that a container is being "watched", and should be reinspected.

The current approach is basically a more manual approach to 2, but I'm thinking about a way to better integrate it into the relist loop.

File: pkg/kubelet/qos/policy.go (hunk 6 above)
Comment (tallclair) [reply pkg/kubelet/qos/policy.go:66]: Yeah, I started making the type change out of curiosity, and it ballooned into a big change very quickly. I might still try to do it after all the other beta changes go in.

File: pkg/kubelet/kubelet.go [hunk 13]
```diff
@@ -1568,7 +1568,7 @@ func (kl *Kubelet) initializeRuntimeDependentModules() {
 		os.Exit(1)
 	}
 	// eviction manager must start after cadvisor because it needs to know if the container runtime has a dedicated imagefs
-	kl.evictionManager.Start(kl.StatsProvider, kl.GetActivePods, kl.PodIsFinished, evictionMonitoringPeriod)
+	kl.evictionManager.Start(kl.StatsProvider, kl.getAllocatedPods, kl.PodIsFinished, evictionMonitoringPeriod)
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/kubelet/kubelet.go:1571]: Done.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet.go]: Yeah not saying we need to add this to the function itself. I do think adding some comments at the call sites would help. Not a blocker though.

File: pkg/kubelet/kubelet.go [hunk 14]
```diff
@@ -2754,117 +2736,87 @@ func (kl *Kubelet) HandlePodSyncs(pods []*v1.Pod) {
...
+		if err := kl.statusManager.SetPodResizeStatus(allocatedPod.UID, resizeStatus); err != nil {
 			//TODO(vinaykul,InPlacePodVerticalScaling): Can we recover from this in some way? Investigate
-			klog.ErrorS(err, "SetPodResizeStatus failed", "pod", klog.KObj(updatedPod))
-			return pod
+			klog.ErrorS(err, "SetPodResizeStatus failed", "pod", klog.KObj(allocatedPod))
 		}
-		updatedPod.Status.Resize = resizeStatus
 	}
-	kl.podManager.UpdatePod(updatedPod)
-	kl.statusManager.SetPodStatus(updatedPod, updatedPod.Status)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go:2866]: Why was this needed previously?

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet.go]: > The first part of the function gets the allocated pod, but if there's no resize (no update), it just returns early with that. Is that what you mean?

My bad. That was a more general question. When kubelet sees a pod for the first time with the desired spec (and no resize request yet), do we also set the allocated resources in the status? 

(BTW, this question doesn't need to block the PR)

File: pkg/kubelet/kubelet.go (hunk 14 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go:2866]: Allocated resources were read from the pod status. Honestly it was very confusing, and there are a lot of bugs around inconsistent use of allocated vs desired. I don't fully understand what the implications of setting these were, but I'm confident they're no longer needed.

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (tallclair) [reply pkg/kubelet/kubelet.go]: The allocated resources are just the admitted resources. So if a new pod is admitted but doesn't have a resize request, then we still set its allocated resources even though they match the spec. Same goes for the container status (although there are a couple changes to the way that API is handled, which I'm tackling in a follow-up PR)

File: pkg/kubelet/kubelet.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vinaykul) [reply pkg/kubelet/kubelet.go:2778]: Yes. I don't know if API changes can be made after Beta (vaguely recall reading that it is allowed during alpha)

**Comment by tooptoop4** [comment]
any ideas for https://github.com/kubernetes/kubernetes/issues/124308
